main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2018 Magnus Madsen, Anna Krogh, Patrick Lundvig, Christian Bonde
+ * 2025 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.optimizer
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur.*
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.{DefContext, Expr, Occur}
+import ca.uwaterloo.flix.language.ast.{AtomicOp, OccurrenceAst, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.util.collection.CofiniteSet
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+import java.util.concurrent.ConcurrentLinkedQueue
+
+/**
+  * The inliner optionally performs beta-reduction at call-sites.
+  */
+object Inliner {
```

Comment:
Docs:

- Closure elimination
- Dead-code removal
- Inline-expansion of defs (and soon linear handlers)

Reply:
Maybe also constant folding? See comments below

Reply:
Examples